Greater usability of a VersionsCollection

nikolaposa opened the pull request:
Changes summary:
- Add `VersionsCollection::isEmpty()` method
- Add `VersionsCollection::toArray()` method
- Add immutable `VersionsCollection::sortedAscending()` and `VersionsCollection::sortedDescending()` methods; mark `VersionsCollection::sort()` as deprecated
- Add `VersionsCollection::first()` method
- Add `VersionsCollection::last()` method
- Add `VersionsCollection::majorReleases()` method
- Add `VersionsCollection::minorReleases()` method
- Add `VersionsCollection::patchReleases()` method

nikolaposa committed:
Add `VersionsCollection::isEmpty()` method.

nikolaposa committed:
Add `VersionsCollection::fromStrings()` named constructor.

nikolaposa committed:
Add `VersionsCollection::toArray()` method.

nikolaposa committed:
Improve VersionsCollection tests.

nikolaposa committed:
Fix coding standard violations.

nikolaposa commented:
@ntzm You might be interested to review/comment this one as well.

ntzm commented:
Looks good to me, thanks for sorting! 👍

nikolaposa commented:
@ntzm Are there any other particular collection features that you might find useful, i.e. for Roave/BackwardCompatibilityCheck library where you're making some contributions? 

Sorry for bothering you with these questions, but the only real feedback I got was from your side.

ntzm commented:
ping @asgrim @Ocramius

Ocramius commented on src/VersionsCollection.php:
```diff
@@ -28,11 +28,23 @@ public function __construct(Version ...$versions)
         return count($this->versions);
     }
 
+    public function isEmpty() : bool
```
Just `empty()`?

Ocramius commented on src/VersionsCollection.php:
```diff
@@ -60,4 +72,9 @@ function (Version $version) use ($constraint) {
         ));
     }
+
+    public function toArray() : array
```
Please document this return type

Ocramius commented on src/VersionsCollection.php:
```diff
@@ -28,11 +28,23 @@ public function __construct(Version ...$versions)
         $this->versions = $versions;
     }
 
+    public static function fromStrings(string ...$versionStrings) : VersionsCollection
```
Wouldn't add this, tbh. New runtime exceptions rather than just a typed ctor

nikolaposa replied:
The only reason why I wanted to restore this feature was that I saw a lot of use cases in Roave/BackwardCompatibilityCheck where it might come handy, though those were mostly in tests. But I agree that type safety should not be sacrificed here. Removing it. /cc @ntzm

Ocramius commented on src/VersionsCollection.php:
```diff
@@ -28,11 +28,23 @@ public function __construct(Version ...$versions)
         $this->versions = $versions;
     }
 
+    public static function fromStrings(string ...$versionStrings) : VersionsCollection
```
Wouldn't add this, tbh. New runtime exceptions rather than just a typed ctor

nikolaposa committed:
Improve VersionCollection::isEmpty() check.

nikolaposa committed:
Document VersionCollection::toArray() for a more precise return type information.

nikolaposa committed:
Remove VersionCollection::fromStrings().

nikolaposa commented:
@Ocramius Can you elaborate this one further: https://github.com/Roave/BackwardCompatibilityCheck/pull/61/files#r186212959? I assume you're suggesting that `sort()` should be returning a new collection (immutability). I wish I had foreseen this earlier, because such a change at this stage would be a BC break.

Also, if I remember well PHPStan detected some issue with the way author tried to use `VersionsCollection`. Perhaps I can extend it with some `first()` and `last()` helper methods, but I would rather try keeping it minimalistic without exposing too much unnecessary staff.

Ocramius commented:
@nikolaposa an improvement could be to make a new `sorted()` method, and make that immutable. 

Then deprecate the old one and remove it in the next major.

Changing the existing method is too implicit and quite problematic.

nikolaposa committed:
Add immutable VersionsCollection::sorted() method to be used instead of sort().

nikolaposa committed:
Add VersionsCollection::first() and VersionsCollection::last() methods.

nikolaposa commented:
@Ocramius I added immutable `VersionsCollection::sorted()` method.

I also added `first()` and `last()` methods, but I'm in a doubt whether I should use nullable return type as it is now or raise an exception instead: https://github.com/nikolaposa/version/pull/18/files#diff-b701aaf410484f65ecf6e06b7f9a411eR41.

Ocramius commented:
> as it is now or raise an exception instead

`first` and `last` can match if the collection is never empty

nikolaposa commented:
Yeah, but the collection can be empty: https://github.com/nikolaposa/version/blob/master/src/VersionsCollection.php#L26. Also, empty collection can be produced after filtering via `matching()`.

Ocramius commented:
@nikolaposa makes sense. A `sorted()` empty collection is still an empty collection, while `first()` and `last()` throwing is indeed OK.

Ocramius commented:
@nikolaposa we should probably talk about this lib over a rakia this week :-)

nikolaposa commented:
@Ocramius LOL, agreed. :)

Ocramius commented:
@nikolaposa as discussed in person, it would be good to have a simplified API dedicated to:

 * people writing version comparison software (this project)
 * people writing release software (next major, next patch, next minor, or figuring out previous versions to create changelogs)

As an example, the API could be similar to:

```php
        $versions
            ->sortedAscending()
            ->previousPatchFor($version);
```

nikolaposa committed:
Add VersionsCollection::sortedAscending() and VersionsCollection::sortedDescending() methods.

nikolaposa committed:
Raise exception when invoking VersionsCollection::first() and VersionsCollection::last() methods on an empty collection.

nikolaposa committed:
Add VersionsCollection::filterMinor() and VersionsCollection::filterPatch() methods.

nikolaposa commented:
@Ocramius I was examining different use cases in BackwardCompatibility library, `PickLastMinorVersionFromCollection` implementation in particular, which gave me the idea for adding `VersionsCollection::filterMinor()` and `VersionsCollection::filterPatch()` methods. Picking last minor version from a collection can now be as simple as:
```
$lastMinor = $versions
    ->filterMinor()
    ->sortedDescending()
    ->first();
```

Ocramius commented:
@nikolaposa `filterMinor()` is not that descriptive: what does it do in your mind? Can you write an explanation for it, that we can then distill to a ubiquitous method name? Same for `filterPatch()`

nikolaposa committed:
Fix VersionsCollection test.

nikolaposa commented:
@Ocramius Yeah, I was struggling to come up with a good name. So for for instance, `filterPatch()` filters-out everything except patch versions (x.y.Z > 0), as in: https://github.com/nikolaposa/version/blob/improvement/versions-collection-greater-usability/tests/VersionsCollectionTest.php#L331.

Do you think something like `filterPatchVersions()`, `findPatchVersions()` or `pickPatchVersions()` describe it better?

Ocramius commented:
> Do you think something like `filterPatchVersions()`, `findPatchVersions()` or `pickPatchVersions()` describe it better?

Not really, and also, those don't seem to be really that useful... Check https://github.com/Roave/BackwardCompatibilityCheck/blob/6c797c5118e7b081eff97db01bcc62f35aa9daf8/src/Git/PickLastMinorVersionFromCollection.php for a fixed version of that consumer.

Ocramius commented:
Some more useful API endpoints:

 * `findLatestMajorRelease()`: given `1.0.0, 2.0.0, 2.1.0`, gives you `2.0.0`
 * `findLatestMinorRelease()`: given `1.0.0, 2.0.0, 2.2.0, 2.2.1`, gives you `2.2.0`
 * `findLatestRelease()`: given `1.0.0, 2.0.0, 2.2.0, 2.2.1`, gives you `2.2.1`

These make more sense for that *specific* use-case, for example.

nikolaposa commented:
I try to avoid methods returning `Version` object, because things become complicated when I need to figure out what to do in case of exceptional conditions which is when nothing is found in this particular case. Returning `null` is always a bad idea in my opinion. Introducing a Special Case object at this stage doesn't make much sense. Raising an exception feels kinda inappropriate for the semantics of routines that "try to find" something.

That's why I wanted to keep `VersionsCollection` consistent and broad in sense that it always return new collection for filtering operations, and let the consumer invoke `isEmpty()`, `first()`, etc. on it:
`findMajorReleases()->sortedDescending()->first()`.

Or, those type of operations could be delegated to some `ReleaseResolver` or `ReleaseFinder` class for better separation of concerns, but I'd rather avoid introducing a new component in this milestone.

I think I'll go with extending `VersionsCollection` API according to your suggestions, but I'll just need to figure out the most meaningful solution for handling exceptional conditions. Ideas are welcome. @Ocramius

Ocramius commented:
@nikolaposa using `->first()` and `->last()` in combination with filters is a very good approach, and I totally support that, but the filtering methods really need better names :D

nikolaposa committed:
Implement VersionsCollection::(major|minor|patch)Releases() methods.

nikolaposa committed:
Fix README example.

nikolaposa commented:
@Ocramius I eventually ended up with:
```
$latestMajorRelease = $releases
    ->majorReleases()
    ->sortedDescending()
    ->first();
```
Also added test case that covers it: https://github.com/nikolaposa/version/pull/18/files#diff-f83b2394fa806ea8e550f37c251c7f76R375

`majorReleases()` still might not be an ideal method name, but perhaps tomorrow I'll have more inspiration for naming. :)

nikolaposa committed:
Move release type checks into Version class.

Ocramius commented:
* `->majorReleases()`

So this basically fetches all `X.0.0`, right? That sounds quite understandable, thanks! :+1:

nikolaposa commented:
Yes, exactly. Thanks for the feedback!

I think this is it as for this PR, but I'll wait with merging to see what @asgrim was up to with: https://twitter.com/asgrim/status/1003608706687397888

Ocramius commented on tests/TestAsset/VersionIsIdentical.php:
```diff
@@ -34,16 +34,16 @@ public function __construct(
+    protected function matches($version) : bool
     {
-        /* @var $constraint Version */
+        /* @var $version Version */
```
Why is the type not constrained in the signature?

nikolaposa replied:
@Ocramius `Declaration should be compatible with Constraint->matches()` warning.

asgrim reviewed (commented):
:+1:
